### PR TITLE
Add support for native binary ops 

### PR DESCRIPTION
**Pull request by DenisVieriu97**
Adds support for native binary kernels, handling both broadcasting, strided tensors / tensors with an offset.

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/BinaryOps.mm [hunk 1]
```diff
@@ -22,6 +24,295 @@
...
+  }
+
+  const Tensor& outputTensor = iter.tensor(0);
+
+  // std::cout << iter.dtype() << std::endl;
+  // std::cout << "Input dtype: " << inputTensor.dtype() << std::endl;
+  // std::cout << "Other dtype: " << otherTensor.dtype() << std::endl;
+  // std::cout << "output dtype: " << outputTensor.dtype() << std::endl;
+  // std::cout << "COMMON DTYPE:" << iter.common_dtype() << std::endl;
+  // std::cout << "ITER SCALAR(0): "<< iter.is_scalar(0) << " " << iter.is_cpu_scalar(0) << std::endl;
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/BinaryOps.mm:185]: cleanup.

File: aten/src/ATen/native/mps/operations/BinaryOps.mm [hunk 2]
```diff
@@ -22,6 +24,295 @@
...
+  } else {
+    inputTensor = iter.tensor(1);
+    otherTensor = iter.tensor(2);
+  }
+
+  if (inputTensor.numel() == 0 || otherTensor.numel() == 0) {
+    return;
+  }
+
+  if (inputTensor.scalar_type() == kDouble) {
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/BinaryOps.mm:171]: Why is double dtype coming at this point ?

File: aten/src/ATen/native/mps/operations/BinaryOps.mm [hunk 3]
```diff
@@ -22,6 +24,295 @@
...
+      }
+
+      for (const auto i: c10::irange(nDim)) {
+        for (const auto offset: c10::irange(nOffsets)) {
+            strides[i][offset] = iter.strides(offset)[i];
+        }
+      }
+
+      id<MTLBuffer> kernelDataOffsets;
+      if (!iter.is_contiguous()) {
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/BinaryOps.mm:237]: Can we unify and cleanup this code with the other fmax/fmin binary kernels ?

File: aten/src/ATen/native/mps/operations/BinaryOps.mm [hunk 4]
```diff
@@ -22,6 +24,295 @@
 typedef MPSGraphTensor* (^BinaryOpBlock)(BinaryOpCachedGraph*, MPSGraphTensor*, MPSGraphTensor*);
 #define BinaryOpFn(graph, primary, secondary) MPSGraphTensor* (mps::BinaryOpCachedGraph* graph, MPSGraphTensor* primary, MPSGraphTensor* secondary)
 
+enum class BinaryKernelType {
```
Comment (razarmehr) [inline aten/src/ATen/native/mps/operations/BinaryOps.mm:27]: Can you please move this entire code to a new file like BinaryOpsKernels.mm? 
The BinaryOps.mm is a mess already, and we don't want to make it worse.

**Review by razarmehr — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mps/operations/BinaryOps.mm [hunk 5]
```diff
@@ -52,7 +617,47 @@ void binaryOpTensor(const Tensor& self, const Tensor& other, const Scalar& alpha
...
+      kernel = "sub";
+    } else if (op_name == "lessThan") {
+      kernel = "less";
+    } else if (op_name == "greaterThan") {
+      kernel = "greater";
+    } else if (op_name == "greaterThanOrEqualTo") {
+      kernel = "greaterOrEqual";
+    } 
+    else {
+      std::cout << op_name << std::endl;
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/BinaryOps.mm:640]: Cleanup.

**Review by kulinseth — APPROVED** [review]

**Comment by razarmehr** [comment]
Failure in test_mps

File: aten/src/ATen/native/mps/operations/BinaryKernel.mm [hunk 6]
```diff
@@ -1,10 +1,387 @@
...
+    if (iter.is_cpu_scalar(2)) {
+      scalar = mps::getMPSScalar(otherTensor.item(), otherTensor.scalar_type());
+      otherBuffer = (id<MTLBuffer>)getIMPSAllocator()->allocScalarBufferWithValue(&scalar.value, scalar.size).get();
+      otherTensorStorage = 0;
+    }
+  } else if (scalar_pos) {
+    type = allContiguous ? BinaryKernelType::RHS_Scalar : BinaryKernelType::Strided_RHS_Scalar;
+    if (iter.is_cpu_scalar(scalar_pos)) {
+      scalar = mps::getMPSScalar(otherTensor.item(), otherTensor.scalar_type());
+      otherBuffer = (id<MTLBuffer>)getIMPSAllocator()->allocScalarBufferWithValue(&scalar.value, scalar.size).get();
```
Comment (razarmehr) [inline aten/src/ATen/native/mps/operations/BinaryKernel.mm:241]: This will crash on x86 as it works only on Unified memory (see getMPSGraphTensorFromScalar()).
We can fix later if you think x86 isn't a priority.

File: aten/src/ATen/native/mps/operations/BinaryKernel.mm [hunk 7]
```diff
@@ -1,10 +1,387 @@
...
+        kernel += "_strided";
+      }
+
+      if (allContiguous) {
+        // Round up to next multiple of 4
+        numThreads = (numThreads + (4 - numThreads % 4)) / 4;
+        gridSize = MTLSizeMake(numThreads, 1, 1);
+      }
+
+      id<MTLComputePipelineState> binaryPSO = mps::getBinaryPSO(device,
```
Comment (razarmehr) [inline aten/src/ATen/native/mps/operations/BinaryKernel.mm:299]: If we want this kernel to appear in Profiler, you need to include `MPSProfiler.h` header and add this after getting your PSO (line 308):
`getMPSProfiler().beginProfileKernel(binaryPSO, kernel, {inputTensor, otherTensor});`

File: aten/src/ATen/native/mps/operations/BinaryKernel.mm [hunk 8]
```diff
@@ -1,10 +1,387 @@
...
+
+      NSUInteger tgSize = binaryPSO.maxTotalThreadsPerThreadgroup;
+      if (tgSize > numThreads) {
+          tgSize = numThreads;
+      }
+
+      MTLSize threadGroupSize = MTLSizeMake(tgSize, 1, 1);
+      [computeEncoder dispatchThreads: gridSize
+                threadsPerThreadgroup: threadGroupSize];
+    }
```
Comment (razarmehr) [inline aten/src/ATen/native/mps/operations/BinaryKernel.mm:326]: If we want Profiler AND the Adaptive Commit for this kernel, you need to add this line here:
`mpsStream->commitAdaptive({inputTensor, otherTensor, outputTensor}, binaryPSO);`

File: aten/src/ATen/native/mps/operations/BinaryKernel.mm [hunk 9]
```diff
@@ -1,10 +1,387 @@
...
+  return true;
+}
+
+bool dispatchNativeBinaryKernel(const Tensor& self,
+                          const Tensor& other,
+                          const Tensor& output,
+                          const Scalar& alpha,
+                          const std::string& op_name) {
+  if (alpha.toFloat() == 1.0   &&
+     (!self.is_contiguous()    ||
```
Comment (razarmehr) [inline aten/src/ATen/native/mps/operations/BinaryKernel.mm:368]: Since we have the check `tensor.is_contiguous() || tensor.storage_offset()` in various places, lets remember to refactor this as a function in OperartionUtils.mm to reuse. We can do this in a separate PR,

**Review by kulinseth — APPROVED** [review]

File: aten/src/ATen/native/mps/operations/BinaryKernel.mm (hunk 7 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/BinaryKernel.mm:358]: Thanks Ramin - yes, I'd like these to appear in the profiler. I'll update this
